### ISSUE TITLE
"del" key deletes backwards

### ISSUE THREAD

**Issue by dxlr8r** [fish-shell/fish-shell#3899]
```
# fish --version
2.5.0
```

Cursor is "|"

Let's say you have prompt: 

`echo |`

And now press delete, nothing happens as it should. 

Then let's say you have: 

`echo |foobar`

Pressing del will now delete "foobar" as it should, but if you continue to press it it will also delete "echo ". 

Tested on macOS Sierra with Terminal, iTerm and xterm. Locally and to RHEL 7 with two accounts (my personal, and one vanilla fish config test account).

**Comment by faho** [comment]
The issue here is probably that there's an ancient argument over what delete/backspace should send.

Often "backspace" will actually send the ASCII DEL (\x7f) character.

What is your $TERM? Are you setting it in e.g. config.fish (if so, please stop doing that)?

What does `fish_key_reader` say when you press those keys? What does `bind` then report those things are bound to? For me backspace sends \x7f, which is bound to "backward-delete-char", while delete sends \c\[3~, which terminfo says is the "dc" key sequence, and which is bound to "delete-char".

**Comment by dxlr8r** [comment]
Bash doesn't have this issue. 

I do set the $TERM from config, using xterm, I've had best experience with that. What is the alternative? 

Output from fish_key_reader with "del" key as input. 

```
[root@ifixit] ~ # fish_key_reader
16:37:55 (2017-03-16)
Press a key

              hex:   1B  char: \c[  (or \e)
(  0.128 ms)  hex:   5B  char: [
(  0.110 ms)  hex:   33  char: 3
(  0.107 ms)  hex:   7E  char: ~
bind -k dc 'do something'
bind \e\[3~ 'do something'
```

Doesn't look like fish_key_reader was installed on my Mac, so this was tested through ssh to a RHEL 7 host. But both have the same behavior.

**Comment by faho** [comment]
>I do set the $TERM from config, using xterm, I've had best experience with that. What is the alternative?

The absolute best thing to do is to let your terminal handle setting $TERM or alternatively set the correct value _in the terminal's configuration_ (slight tweaks are okay here, e.g. some terms on some OSs still default to `xterm` when they can do `xterm-256color`). If neither is possible, switch to a terminal that allows it. Terminals behave differently, and without $TERM there's no way for us to tell which way it'll behave.

>Output from fish_key_reader with "del" key as input.

So you pressed "del" and got "dc"? That would be correct. What does `bind -k dc` say?

------------------------


After actually reading the OP in its entirety (you know, like we tell kids in elementary school - maybe I should go there again) and testing it, I can confirm all this. I still strongly advise you to not set $TERM manually, but there is an actual issue here.

**Comment by dxlr8r** [comment]
```
# bind -k dc
bind -k dc delete-char
```

I commented out the manually assignment of xterm, same behavior (and it still use xterm, set from iTerm as well). The reason I've done it this way is that where i work we have a lot of different machines connected through bastions (jump hosts, etc.). I do remember there was a reason I had to do it this way, but I'll look into removing it.

**Comment by faho** [comment]
Ha, gotcha! You also didn't read my entire comment!

Seriously: The behavior you described is occuring, and it has nothing to do with misconfiguration. In fact, it doesn't have anything to do with _input_. It's down to the behavior of the "delete-char" bind function, which was changed in e2ed56126438747c2b24d9b1d8ec84788303939c for vi-mode (since that's how "x" works there).

So now we have a couple of options:

- Revert that commit and instead bind vi's "x" to "delete-char; backward-char" (which would also move the cursor at the end of the line, which might actually be better than doing nothing)

- Change the emacs binding to "delete-char; forward-char" (won't work properly at the beginning of the line)

- Make another bind function ("delete-char-and-move"?)

- Leave it as-is

**Comment by jomo** [comment]
Please don't leave it as-is. Whatever you do, please don't move the cursor, it's extremely annoying. No other shell does this, they all delete the character on the right without moving the cursor (in any case)
![fish bug gif](https://cloud.githubusercontent.com/assets/3585743/24408267/65f8eace-13cd-11e7-8319-0d9625ebba4b.gif)

**Comment by gima** [comment]
@faho wrote:
> It's down to the behavior of the "delete-char" bind function, which was changed in e2ed561 for vi-mode (since that's how "x" works there).

So, is the following correct?:  
A change was made in commit e2ed561 that changed the way a built-in function "delete-char" works. The commit was titled (and thus supposed to fix something in) "VI-compatible keybinding mode".

My keybindings are the default, i.e. *not* VI-mode - which I make sure by running the command `fish_default_key_bindings` (whose description says "Default (Emacs-like) key bindings for fish").

Combining this with the fact that when I press the delete key, I expect (and have for over ten years) it do one thing: delete. Not delete *and* move cursor. Therefore I see this very clearly as a bug and a blatant violation of user interface design [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Such basic things as "what does a key on my keyboard do when I press it" should NOT change when a software updates (unless there is a very good reason).

The default key bindings as I see are NOT vi-mode. And since the change (e2ed561) was supposed to affect only vi-mode, it should NOT affect the default keybindings - just the vi-mode

@faho wrote:
> So now we have a couple of options:
> * Revert that commit and instead bind vi's "x" to "delete-char; backward-char" (which would also move the cursor at the end of the line, which might actually be better than doing nothing)

This is exactly what should be done. If a "controversial" change is to be done, make it opt-in for those who want it, not opt-out for those who dont.

<sub>Goddammit justifying this is such a waste of time that could easily have been avoided if people just though for a moment rather than bulldozing ahead with their own agenda! <sup> steam <sup> huff </sup><sup> puff <sup>..</sup></sup></sup></sub>

**Comment by krader1961** [comment]
@gima: We appreciate how annoying issues like this one are. And I know how satisfying it can feel to vent. But your comment is counterproductive. You seem to think we deliberately broke the emacs mode behavior and that we did so without carefully considering how such a change would impact users. Neither is true. You also say "I see this very clearly as a bug". And so do we. The issue is clearly labeled a bug. Your comment only serves to demotivate people, like myself, to fix it. 

If you feel the need to rant you should also be prepared to take ownership of the issue and do the work to fix it. Otherwise you might want to take a deep breath and consider using more constructive phrasing before pressing the [comment] button.

**Comment by faho** [comment]
Please everyone test #3969.

**Comment by krader1961** [comment]
Ideally we would fix this in the 2.6.0 release which is already a few days past its target date. But unless @dxlr8r or @gima or someone else who feels passionately about this verifies that the proposed fix by @faho is acceptable it will have to slip to a subsequent release.
